src/neo/SmartContract/ApplicationEngine.cs

```diff
@@ -134,6 +135,14 @@ protected ApplicationEngine(TriggerType trigger, IVerifiable container, DataCach
             this.gas_amount = gas;
             this.exec_fee_factor = snapshot is null || persistingBlock?.Index == 0 ? PolicyContract.DefaultExecFeeFactor : NativeContract.Policy.GetExecFeeFactor(Snapshot);
             this.StoragePrice = snapshot is null || persistingBlock?.Index == 0 ? PolicyContract.DefaultStoragePrice : NativeContract.Policy.GetStoragePrice(Snapshot);
+            this.nonceData = container is Transaction tx ? tx.Hash.ToArray()[..16] : new byte[16];
+            if (persistingBlock is not null)
+            {
+                fixed (byte* p = nonceData)
+                {
+                    *(ulong*)p ^= persistingBlock.Nonce;
```

Comment:
When it's a block it has 8 bytes, and 8 zeros, but if it's temporal, I am not against to it